graph/hopcroft_karp.cpp

```diff
@@ -0,0 +1,291 @@
+/**
+ * @file 
+ * @brief  Implementation of [Hopcroft–Karp](https://en.wikipedia.org/wiki/Hopcroft%E2%80%93Karp_algorithm) algorithm.
+ * @details 
+ * The Hopcroft–Karp algorithm is an algorithm that takes as input a bipartite graph 
+ * and produces as output a maximum cardinality matching, it runs in O(E√V) time in worst case.
+ * 
+ * ### Bipartite graph
+ * A bipartite graph (or bigraph) is a graph whose vertices can be divided into two disjoint 
+ * and independent sets U and V such that every edge connects a vertex in U to one in V. 
+ * Vertex sets U and V are usually called the parts of the graph. 
+ * Equivalently, a bipartite graph is a graph that does not contain any odd-length cycles.
+ * 
+ * ### Matching and Not-Matching edges
+ * Given a matching M, edges that are part of matching are called Matching edges and edges that are not part 
+ * of M (or connect free nodes) are called Not-Matching edges.
+ * 
+ * ### Maximum cardinality matching
+ * Given a bipartite graphs G = ( V = ( X , Y ) , E ) whose partition has the parts X and Y, 
+ * with E denoting the edges of the graph, the goal is to find a matching with as many edges as possible. 
+ * Equivalently, a matching that covers as many vertices as possible.
+ * 
+ * ### Augmenting paths
+ * Given a matching M, an augmenting path is an alternating path that starts from and ends on free vertices. 
+ * All single edge paths that start and end with free vertices are augmenting paths.
+ * 
+ * 
+ * ### Concept
+ * A matching M is not maximum if there exists an augmenting path. It is also true other way,
+ * i.e, a matching is maximum if no augmenting path exists.
+ * 
+ * 
+ * ### Algorithm
+ * 1) Initialize the Maximal Matching M as empty.
+ * 2) While there exists an Augmenting Path P
+ *   Remove matching edges of P from M and add not-matching edges of P to M
+ *   (This increases size of M by 1 as P starts and ends with a free vertex
+ *   i.e. a node that is not part of matching.)
+ * 3) Return M. 
+ * 
+ * 
+ *
+ * @author [Krishna Pal Deora](https://github.com/Krishnapal4050)
+ * 
+ */
+
+
+#include <iostream>
+#include <cstdlib> 
+#include <queue>
+#include <list>
+#include <climits>
+#include <memory>
+#include <cassert>
+
+/**
+ * @namespace graph 
+ * @brief Graph algorithms
+ */
+ namespace graph { }
```

Comment:
```suggestion
 namespace graph { 
```
only the opening parentheses

Reply:
@kvedala If I will keep only the opening parentheses, then the code formatter will show error because there must be a closing parentheses for it.

Reply:
You are correct. In my previous review, I provided that as well. https://github.com/TheAlgorithms/C-Plus-Plus/pull/1087#discussion_r498324142

Reply:
There are other suggestions made in my review that you have marked as un-resolved without reviewing. Please recheck them. Thank you

Reply:
@kvedala I have corrected the namespace parentheses as suggested by you. Thanks a lot !

Reply:
@kvedala I have also added two new sample test cases. Please review it once